plugins/nominal-connection-checker/src/priority_queue_map.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Defines the PriorityQueueMap and its helper prototypes.
+ */
+'use strict';
+
+/**
+ * A map where keys can be bound to multiple values, which are given a priority.
+ * When you get the value of a key it returns all of the values with the highest
+ * priority.
+ */
+export class PriorityQueueMap {
+  /**
+   * Constructs the PriorityQueueMap.
+   */
+  constructor() {
+    /**
+     * Maps the keys to an array of Bindings.
+     * @type {!Object<*, !Array<Binding>>}
+     * @private
+     */
+    this.map_ = Object.create(null);
+  }
+
+  /**
+   * Returns the highest priority values for the given key. Or undefined if the
+   * key is not bound.
+   * @param {*} key The key to find the value of.
+   * @return {undefined|!Array<*>} The highest priority values for the given
+   *     key. Or undefined if the key is not bound.
+   */
+  getValues(key) {
+    const bindings = this.getBindings(key);
+    return bindings && bindings.map((binding) => binding.value);
+  }
+
+  /**
+   * Returns the highest priority Bindings for the given key. Or undefined if
+   * the key is not bound.
+   * @param {*} key The key to find the binding of.
+   * @return {undefined|!Array<!Binding>} The highest priority Bindings for the
+   *     given key. Or undefined if the key is not bound.
+   */
+  getBindings(key) {
+    const bindings = this.map_[key];
+    if (!bindings || !bindings.length) {
+      return undefined;
+    }
+    let highestBindings = [bindings[0]];
+    let highestPriority = bindings[0].priority;
+    for (let i = 1, binding; (binding = bindings[i]); i++) {
+      if (binding.priority > highestPriority) {
+        highestPriority = binding.priority;
+        highestBindings = [binding];
+      } else if (binding.priority == highestPriority) {
+        highestBindings.push(binding);
+      }
+    }
+    return highestBindings;
+  }
+
+  /**
+   * Adds the given value with the associated priority to the key's priority
+   * queue of values.
+   * @param {*} key The key to bind the value to.
+   * @param {*} value The value to bind to the key.
+   * @param {number} priority The priority of the binding.
+   */
+  bind(key, value, priority) {
+    if (!this.map_[key]) {
+      this.map_[key] = [];
+    }
+    this.map_[key].push(new Binding(value, priority));
+  }
+
+  /**
+   * Removes the given value with the associated priority from the key's
+   * priority queue of values.
+   * @param {*} key The key to unbind the value from.
+   * @param {*} value The value to unbind from the key.
+   * @param {number} priority The priority of the binding.
+   */
+  unbind(key, value, priority) {
```

Comment:
Is it possible to have multiple values with the same priority?

Reply:
It is possible to have:
* Multiple different values with different priorities.
* Multiple different values with the same priority.
* Multiple of the same value with different priority.
* Multiple of the same value with the same priority.

So basically: Yes. The PriorityQueueMap doesn't really have any fancy behavior atm, it's just a wrapper for a map where all of the values are priority queues.